[WIP] Wrap C++17 std::filesystem::path to solve encoding issue on Windows.

ken2812221 opened the pull request:
Fix #13103
We can call `.u8string()` everywhere and call `.string()` when we need to use native-encoded path.

ken2812221 committed:
[WIP] Wrap C++17 std::filesystem::path to solve encoding issue on Windows

laanwj commented:
Neat.
Concept ACK!

fanquake commented:
Travis macOS build failed with:
```
In file included from util.cpp:6:
In file included from ./util.h:18:
In file included from ./fs.h:11:
In file included from /home/travis/build/bitcoin/bitcoin/depends/x86_64-apple-darwin11/share/../include/boost/filesystem.hpp:16:
/home/travis/build/bitcoin/bitcoin/depends/x86_64-apple-darwin11/share/../include/boost/filesystem/path.hpp:141:33: error: no type named 'type' in 'boost::enable_if<boost::filesystem::path_traits::is_pathable<Path>, void>'; 'enable_if' cannot be used to disable this declaration
      typename boost::enable_if<path_traits::is_pathable<
                                ^~~~~~~~~~~~~~~~~~~~~~~~~
./util.h:315:36: note: in instantiation of member function 'boost::filesystem::path::path' requested here
    using boost::filesystem::path::path;
                                   ^
util.cpp:1098:12: note: while substituting deduced template arguments into function template 'Path' [with Source = Path]
    return Path(Utf8ToNative(source));
           ^
1 error generated.
```

laanwj commented on src/util.h:
```diff
@@ -309,4 +309,13 @@ std::unique_ptr<T> MakeUnique(Args&&... args)
  */
 int ScheduleBatchPriority(void);
 
+class Path : public boost::filesystem::path
```
Would be better in `src/fs.h`?

laanwj commented on src/util.cpp:
```diff
@@ -1020,3 +1022,45 @@ int ScheduleBatchPriority(void)
 #endif
 }
+
+static std::string NativeToUtf8(const std::string& source)
```
string encoding should probably be in `utilstrencodings.[ch]`

laanwj commented:
> Fix #13103

How does this fix that issue? 
This adds a class and some functions, but never uses them, or am I missing something?

> We can call .u8string() everywhere and call .string() when we need to use native-encoded path.

Concept ACK, though I'm not entirely convinced on wrapping versus using functions where needed.

ken2812221 commented:
I'm trying to re-do #13107 with pretty code, but it's still WIP. The target is to fix #13103

ken2812221 committed:
scripted-diff: Use fsbridge::Path instead of boost::filesystem::path

-BEGIN VERIFY SCRIPT-
sed -i 's/fs::path/fsbridge::Path/g' $(git grep --name-only 'fs::path' -- '*.cpp' '*.h' | grep -v 'fs.h' | grep -v 'fs.cpp')
sed -i 's/boost::filesystem::path/fsbridge::Path/g' $(git grep --name-only 'boost::filesystem::path' -- '*.cpp' '*.h' | grep -v 'fs.h' | grep -v 'fs.cpp')
-END VERIFY SCRIPT-

ken2812221 committed:
Move fs::path(string) to fsbridge::U8Path(string)

ken2812221 committed:
Read command line arguments as utf8

ken2812221 committed:
scripted-diff: Use `.u8string()` instead of `.string()`

-BEGIN VERIFY SCRIPT-
sed -i 's/\.string()/\.u8string()/g' $(git grep --name-only '\.string()' -- '*.cpp' '*.h' | grep -v 'fs.h' | grep -v 'fs.cpp')
-END VERIFY SCRIPT-

ken2812221 commented:
Seems not possible to wrap all filesystem classes. Close it and reopen #13107
